python/src/uagents/context.py

```diff
@@ -329,249 +336,136 @@ def get_agents_by_protocol(
-            timeout (Optional[int]): The optional timeout for sending the message, in seconds.
+            schema_digest (str): The schema digest of the message to check.
 
         Returns:
-            MsgStatus: The delivery status of the message.
+            bool: Whether the message is a valid interval message.
         """
-        self._session = self._session or uuid.uuid4()
+        if self._interval_messages:
+            return schema_digest in self._interval_messages
+        return True
 
-        # Check if this message is a reply
-        if (
-            self._message_received is not None
-            and self._replies
-            and schema_digest != ERROR_MESSAGE_DIGEST
-        ):
-            received = self._message_received
-            # Ensure the reply is valid
-            if (received.schema_digest in self._replies) and (
-                schema_digest not in self._replies[received.schema_digest]
-            ):
-                self._logger.exception(
-                    f"Outgoing message {message_type or ''} "
-                    f"is not a valid reply to {received.message}"
-                )
-                return MsgStatus(
-                    status=DeliveryStatus.FAILED,
-                    detail="Invalid reply",
-                    destination=destination,
-                    endpoint="",
-                )
-        # Check if this message is a valid interval message
-        if (
-            self._message_received is None
-            and self._interval_messages
-            and schema_digest not in self._interval_messages
-        ):
-            self._logger.exception(
-                f"Outgoing message {message_type} is not a valid interval message"
-            )
+    async def send(
+        self,
+        destination: str,
+        message: Model,
+        sync: Optional[bool] = False,
+        timeout: Optional[int] = DEFAULT_ENVELOPE_TIMEOUT_SECONDS,
+    ) -> MsgStatus:
+        """
+        This is the pro-active send method which is used in on_event and
+        on_interval methods. In these methods, interval messages are set but
+        we don't have access properties that are only necessary in re-active
+        contexts, like 'replies', 'message_received', or 'protocol'.
+        """
+        schema_digest = Model.build_schema_digest(message)
+        message_body = message.json()
+
+        if not self._is_valid_interval_message(schema_digest):
+            log(self.logger, logging.ERROR, f"Invalid interval message: {message}")
             return MsgStatus(
                 status=DeliveryStatus.FAILED,
                 detail="Invalid interval message",
                 destination=destination,
                 endpoint="",
             )
 
+        return await self.send_raw(
+            destination,
+            schema_digest,
+            message_body,
+            sync=sync,
+            timeout=timeout,
+        )
+
+    async def send_raw(
```

Comment:
If we need to keep this method public, we should also add a docstring explaining when this should be used (or rather why the common developer should never need to use this)